### ISSUE TITLE
[libbpf-cargo] Add ability for user defined Default

### ISSUE THREAD

**Issue by masmullin2000** [libbpf/libbpf-rs#78]
Arrays cannot be #[derive(Default)] if they are > 32 in length.

As such, the line https://github.com/libbpf/libbpf-rs/blob/master/libbpf-cargo/src/btf/btf.rs#L348 will fail compilation for larger arrays.  Because the struct is defined as default per this line of code, the user cannot impl Default themselves.

Arrays of longer than 32 can have user defined Default so long as the generated skeleton does not #[derive(Default)].

Request: Add a flag so that *_bss_types do not automatically add the Default Derive.

Request++: automatically detect if an array in a structure is > 32 length, and create an impl Default in the skeleton

Workaround: cargo libbpf build && cargo libbpf gen && EDIT THE *.skel.rs FILE MANUALLY && add user defined impl Default for types

**Comment by danobi** [comment]
Thanks for the report.

> Request++: automatically detect if an array in a structure is > 32 length, and create an impl Default in the skeleton

I think this is the best approach. I'd rather not add flags when we can just do the right thing. I think it should be straightforward to do this too. I'll take a look tomorrow.

**Comment by masmullin2000** [comment]
@danobi Would you mind if I tried my hand at this ticket?

I would like to become a contributor to this project, and this bug is something I'd like to start with if it's OK with the rest of you.

**Comment by danobi** [comment]
> I would like to become a contributor to this project, and this bug is something I'd like to start with if it's OK with the rest of you.

That would be much appreciated :)

**Comment by danobi** [comment]
Feel free to comment here if you need any hints or have questions

**Comment by mimullin-bbry** [comment]
I have created a "first draft" of the solution.  I have yet to implement proper tests (only one test added), but I'd like feedback to see if I am totally off course or not.

https://github.com/mimullin-bbry/libbpf-rs/commit/bb678ba0cb4aadf08be924e46d2030aaa3d05cf4

**Comment by danobi** [comment]
@mimullin-bbry it looks like you went a step further than what I was thinking and actually generated code for proper `Default` implementations. I was actually just thinking we omit the `#[derive(Default)]` from structs whose members (recursively) contain a  `[T;N]`, where N > 32. Then the user can implement their own Default. What are your thoughts about that?

**Comment by mimullin-bbry** [comment]
I've worked on projects where a developer needed to maintain equivalence between the C-BPF struct and the struct as seen in the 'userspace' codebase (we used c++/bcc).

It was painful and could cause subtle bugs.  I'm not sure if Rust would erase the subtle bugs by just not compiling if the developer defined Default didn't properly match the C-BPF, but the pain of maintaining the structure twice (defining the structure in X.bpf.c and impl Default in whatever the developer.rs file) should be erased if possible.

My reluctance with the code I have written is that it's not as elegant as I would like, but I do think it's important to have the generated impl Default.  I don't like the names I've used, I don't like that I had to introduce a bunch of variables that might get thrown away; but the end result is more preferable to a user impl Default.

One important Note: You may want to avoid this bug/functionality all together.  Eventually const generics will solve the > 32 length default problem
https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta.html

When Rust solves this problem, you'll want to revert back to just #[derive(Default)] for everything.  But that day is not today.

**Comment by danobi** [comment]
> It was painful and could cause subtle bugs. I'm not sure if Rust would erase the subtle bugs by just not compiling if the developer defined Default didn't properly match the C-BPF, but the pain of maintaining the structure twice (defining the structure in X.bpf.c and impl Default in whatever the developer.rs file) should be erased if possible.

I think rust errors out if the Default implementation doesn't match the struct def. That being said, I buy your argument. Probably better to have libbpf-cargo just do the right thing.

> One important Note: You may want to avoid this bug/functionality all together. Eventually const generics will solve the > 32 length default problem
https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta.html

Yeah, was hoping the ongoing work solved this issue but doesn't look like it's done yet.

**Comment by danobi** [comment]
I think the logic might be a bit cleaner if we could have these kind of code paths:

```
if can_use_derive_default(struct) {
  header = "#[derive(Default, ...)]"
} else {
  header = "#[derive(...)]"
}

<the existing generation stuff>

if !can_use_derive_default(struct) {
  emit_manual_default_implementation()
}
```

IOW, I think the logic would be easier to follow if we put stuff in helper methods. It might be easier if you put up a PR so I could comment on the code easier.

**Comment by willfindlay** [comment]
This can probably be closed now since it was fixed in #101

**Comment by masmullin2000** [comment]
fixed
